Implemented RuntimeUser

punkyiang opened the pull request:
- Added RuntimeUser attribute,
- Implemented save/load methods for RuntimeUser,
- Included autoload from disk during init for RuntimeUser

punkyiang committed:
Implemented RuntimeUser

- Added RuntimeUser attribute,
- Implemented save/load methods for RuntimeUser,
- Included autoload from disk during init for RuntimeUser

AntonC9018 commented on Runtime/Settings/Settings.cs:
```diff
@@ -135,15 +146,73 @@ protected void Set<S>( ref S setting, S value )
+
+            try
+            {
+                var filepath = GetSettingsPath() + ( attribute.filename ?? typeof(T).Name ) + ".json";
```
Should check to see if the file exists and return false if it does not, prior to exception handling.

punkyiang replied:
done in latest commit

AntonC9018 commented on Runtime/Settings/Settings.cs:
```diff
@@ -135,15 +146,73 @@ protected void Set<S>( ref S setting, S value )
+        /// Load the current setting from disk, only for RuntimeUser
+        /// </summary>
+        /// <returns>True if load goes successfully, false if errors are present</returns>
+        public static bool LoadSettings()
```
Rename the method to `TryLoadSettings()` (implies bool result, does not throw), rather than `LoadSettings()` (implies throwing on error).

AntonC9018 commented on Runtime/Settings/Settings.cs:
```diff
@@ -135,15 +146,73 @@ protected void Set<S>( ref S setting, S value )
+        {
+            if( attribute.usage != SettingsUsage.RuntimeUser )
+            {
+                Debug.LogWarning( "Load method is only used with RuntimeUser settings" );
```
Slap a `Debug.Assert()` here instead.

AntonC9018 commented on Runtime/Settings/Settings.cs:
```diff
@@ -135,15 +146,73 @@ protected void Set<S>( ref S setting, S value )
+        {
+            if( attribute.usage != SettingsUsage.RuntimeUser )
+            {
+                Debug.LogWarning( "Save method is only used with RuntimeUser settings" );
```
Slap a `Debug.Assert()` here instead.

AntonC9018 commented on Runtime/Settings/Settings.cs:
```diff
@@ -135,15 +146,73 @@ protected void Set<S>( ref S setting, S value )
+
+            if( !ignoreDirtyFlag && !_isDirty )
+            {
+                Debug.Log( "Settings are clean, no save needed" );
```
Pretty much a useless log. I don't think it adds any value.

AntonC9018 commented on Runtime/Settings/Settings.cs:
```diff
@@ -41,7 +42,14 @@ protected static T Initialize()
+                // first load the default setting, then overwrite it
+                _instance = Resources.Load<T>( filename );
+                if (!LoadSettings())
+                    SaveSettings( true );
```
Use a named argument here, `ignoreDirtyFlag : true`. Never have lonely bools unless they don't matter in the given context.

punkyiang committed:
polished Settings.cs

- cached path and filename
- added file existence check in load method
- renamed load method to TryLoadSettings to imply a bool result and no exceptions thrown
- replaced save/load method checks with asserts
- removed redundant logs
